### PR TITLE
Implement different bases (radix) for int#toString

### PR DESCRIPTION
**Pull request by MaxGraey**
this PR introduce:
1) Special fast path for itoa funcs with radix = 16;
2) General path for radix differ than 10 or 16 with fast paths if radix is pow of two.
3) remove generic `itoa<T>`
4) tests

### CODE REVIEW

File: std/assembly/util/number.ts [hunk 1]
```diff
@@ -139,7 +164,22 @@ export function decimalCount64High(value: u64): u32 {
...
+  // while (t <= n) {
+  //   t *= b; e++;
+  // }
+  // return e;
+  var b = u64(base), e: u32 = 1;
+  while (n > 1) {
+    n /= b; e++;
+  }
+  return e;
+}
```
Comment (MaxGraey) [inline std/assembly/util/number.ts:180]: That's hardest and inefficient part for general case.
It's quite hard implement integer logarithm with any base efficiently. It's problem pretty similar to discrete logarithm except we don't care about prime modules.

@jedisct1 may be you suggest something?

I really don't like do twice allocations extra loop for string reversing if we don't pre-calculate number of decimals via ulogBase.

File: std/assembly/util/number.ts [hunk 2]
```diff
@@ -139,7 +164,24 @@ export function decimalCount64High(value: u64): u32 {
   }
 }
 
-function utoa32_lut(buffer: usize, num: u32, offset: usize): void {
+function ulogBase(num: u64, base: i32): u32 {
```
Comment (dcodeIO) [inline std/assembly/util/number.ts:167]: Perhaps `ulog_base` to match surrounding code

File: std/assembly/util/number.ts [hunk 3]
```diff
@@ -203,118 +245,219 @@ function utoa64_lut(buffer: usize, num: u64, offset: usize): void {
...
-  utoa32_lut(buffer, <u32>num, offset);
+  utoa32_dec_lut(buffer, <u32>num, offset);
 }
 
-function utoa_simple<T extends number>(buffer: usize, num: T, offset: usize): void {
+function utoa_hex_lut(buffer: usize, num: u64, offset: usize): void {
+  const lut = changetype<usize>(HEX_DIGITS);
+  while (offset >= 2) {
+    offset -= 2;
+    let digits = load<u32>(lut + ((<usize>num & 0xFF) << alignof<u32>()));
```
Comment (dcodeIO) [inline std/assembly/util/number.ts:255]: Variable seems unnecessary

File: std/assembly/util/number.ts [hunk 4]
```diff
@@ -203,118 +245,219 @@ function utoa64_lut(buffer: usize, num: u64, offset: usize): void {
...
+    throw new RangeError("toString() radix argument must be between 2 and 36");
+  }
+  if (!value) return "0";
+  var out: usize = 0;
+
+  if (radix == 10) {
+    let decimals = decimalCount32(value);
+    out = __alloc(decimals << 1, idof<String>());
+    utoa32_dec_core(out, value, decimals);
+  } else if (radix == 16) {
```
Comment (dcodeIO) [inline std/assembly/util/number.ts:357]: Would it make sense to also handle other common cases, like radix 2 or 8?

File: std/assembly/util/string.ts [hunk 5]
```diff
@@ -865,8 +865,26 @@ export function joinBooleanArray(dataStart: usize, length: i32, separator: strin
 export function joinIntegerArray<T>(dataStart: usize, length: i32, separator: string): string {
   var lastIndex = length - 1;
   if (lastIndex < 0) return "";
-  // @ts-ignore: type
-  if (!lastIndex) return changetype<string>(itoa<T>(load<T>(dataStart))); // retains
+  if (!lastIndex) {
+    let value = load<T>(dataStart);
+    if (isSigned<T>()) {
+      if (sizeof<T>() <= 4) {
```
Comment (dcodeIO) [inline std/assembly/util/string.ts:871]: Good catch!

File: std/assembly/util/number.ts (hunk 4 above)
Comment (MaxGraey) [reply std/assembly/util/number.ts:357]: I don't think so. This bases have rarely usage and usually for debugging / tracing proposes so I don't think we should do special optimizations for them. Btw `utoa64_any_core` already has fast path for pow of two bases like (2, 8, ..., 32) which don't use shifts instead division so it's pretty enough.

**Comment by dcodeIO** [comment]
Didn't check every single detail, but code looks good and tests including the new ones are passing, hence merging :)

**Pull request merged**